vault/diagnose/helpers.go

```diff
@@ -0,0 +1,79 @@
+package diagnose
+
+import (
+	"context"
+	"go.opentelemetry.io/otel"
+	"go.opentelemetry.io/otel/attribute"
+	"go.opentelemetry.io/otel/codes"
+	sdktrace "go.opentelemetry.io/otel/sdk/trace"
+	"go.opentelemetry.io/otel/trace"
+)
+
+const (
+	warningEventName = "warning"
+	actionKey        = "actionKey"
+	errorMessageKey  = attribute.Key("error.message")
+)
+
+var tp *sdktrace.TracerProvider
+var tracer trace.Tracer
+var tc *TelemetryCollector
+
+// Init initializes a Diagnose tracing session.  In particular this wires a TelemetryCollector, which
+// synchronously receives and tracks OpenTelemetry spans in order to provide a tree structure of results
+// when the outermost span ends.
+func Init() {
+	tc = NewTelemetryCollector()
+	//so, _ := stdout.NewExporter(stdout.WithPrettyPrint())
+	tp = sdktrace.NewTracerProvider(
+		sdktrace.WithSampler(sdktrace.AlwaysSample()),
+		//sdktrace.WithSpanProcessor(sdktrace.NewSimpleSpanProcessor(so)),
+		sdktrace.WithSpanProcessor(tc),
+	)
+	otel.SetTracerProvider(tp)
+	tracer = tp.Tracer("vault-diagnose")
+}
```

Comment:
oh i was also meaning to ask, why is this a public `Init()` (that presumably is called somewhere else) vs just using a go `init()`?

Reply:
Because it shouldn't happen all the time, just on runs of diagnose.  It also may need to be more stateful than it currently is, but there are some considerations about the global OpenTelemetry trace provider in that case.